### PR TITLE
ENH: add TabularMSA.sort, from_dict, and to_dict

### LINKED ISSUE #1082: `TabularMSA`: from_dict/to_dict
From [RFC](https://github.com/biocore/scikit-bio-rfcs/blob/master/active/001-tabular-msa.md):

- `from_dict(dict)`

   Return a new TabularMSA from the keys and sequences of a dict. The values
   must all be an appropriate `IUPACSequence` class of the same length.

- `to_dict() -> dict`

   Returns a dictionary that maps the keys of the MSA to the aligned sequences.

### LINKED ISSUE #1095: `TabularMSA`: sort
From [RFC](https://github.com/biocore/scikit-bio-rfcs/blob/master/active/001-tabular-msa.md):

- `sort(key=None, reverse=False) -> None`

   Sort sequences in-place. If `key=None` the existing keys will be used to
   sort. Otherwise `key` matches the constructor. If there are no keys for the
   MSA and `key=None`, an exception should be raised.

### PR DESCRIPTION
**Pull request by jairideout**
Also made ``TabularMSA.keys`` use ``object`` dtype to prevent lossy conversions.

Fixes #1082, fixes #1095.

### CODE REVIEW

File: skbio/alignment/tests/test_tabular_msa.py [hunk 1]
```diff
@@ -493,6 +525,350 @@ def test_reindex_non_hashable_keys(self):
...
+        # original state is maintained
+        self.assertEqual(msa, TabularMSA([]))
+
+        msa = TabularMSA([DNA('TC'), DNA('AA')])
+        with self.assertRaises(OperationError):
+            msa.sort()
+        self.assertEqual(msa, TabularMSA([DNA('TC'), DNA('AA')]))
+
+    def test_sort_on_unorderable_msa_keys(self):
+        # TODO is there a better way to handle this?
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: Couldn't figure out a better way to get this test to pass on Python 2 and 3. @ebolyen any ideas?

File: skbio/alignment/tests/test_tabular_msa.py [hunk 2]
```diff
@@ -493,6 +525,350 @@ def test_reindex_non_hashable_keys(self):
...
+        if future.utils.PY3:
+            msa = TabularMSA([DNA('AAA'), DNA('ACG')], keys=[42, None])
+            with self.assertRaises(TypeError):
+                msa.sort()
+            self.assertEqual(
+                msa,
+                TabularMSA([DNA('AAA'), DNA('ACG')], keys=[42, None]))
+
+    def test_sort_on_unorderable_key(self):
+        # TODO is there a better way to handle this?
```
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: Same goes for this one.

File: skbio/alignment/_tabular_msa.py [hunk 3]
```diff
@@ -564,9 +608,140 @@ def reindex(self, key=None, keys=None):
...
+        Sort the sequences using the MSA's existing keys:
+
+        >>> msa.sort()
+        >>> msa == TabularMSA([DNA('AC-'), DNA('AC-'), DNA('ACG')],
+        ...                   keys=['a', 'b', 'c'])
+        True
+
+        """
+        if key is None:
+            sort_keys = self.keys
```
Comment (ebolyen) [inline skbio/alignment/_tabular_msa.py]: Can you call `tolist()` on `self.keys`?

File: skbio/alignment/tests/test_tabular_msa.py (hunk 1 above)
Comment (ebolyen) [inline skbio/alignment/tests/test_tabular_msa.py]: Try using `object()` it shouldn't be orderable

File: skbio/alignment/_tabular_msa.py (hunk 3 above)
Comment (jairideout) [inline skbio/alignment/_tabular_msa.py]: Yep, fixed here and below

File: skbio/alignment/tests/test_tabular_msa.py (hunk 1 above)
Comment (jairideout) [inline skbio/alignment/tests/test_tabular_msa.py]: `object` is orderable in py2. The following dummy class seems to work though:

```python
class Unorderable(object):
    def __lt__(self, other):
        raise TypeError()
    __cmp__ = __lt__
```

**Comment by jairideout** [comment]
@ebolyen thanks for reviewing! I addressed your comments.

**Comment by ebolyen** [comment]
Thanks @jairideout!

**Pull request merged**